File: 2016/05.cs
using System.Security.Cryptography;
using System.Text;

namespace _2016;

public abstract class _05
{
    private static readonly string Data;

    static _05() => Data = Task.Run(() => Utils.GetInputData(5)).Result.Trim();

	private static string ComputeHash(string doorId, int index)
    {
        var input = doorId + index;
        var hashBytes = MD5.HashData(Encoding.ASCII.GetBytes(input));
        return BitConverter.ToString(hashBytes).Replace("-", "").ToUpper();
    }

    private static string SolvePartOne()
    {
        var password = new StringBuilder();
        var index = 0;

        while (password.Length < 8)
        {
            var hash = ComputeHash(Data, index);

            if (hash.StartsWith("00000")) password.Append(hash[5]);

            index++;
        }
        return password.ToString();
    }

    private static string SolvePartTwo()
    {
        const int passwordLength = 8;
        var password = new char?[passwordLength];
        var filledPositions = 0;
        var index = 0;

        while (filledPositions < passwordLength)
        {
            var hash = ComputeHash(Data, index);

            if (hash.StartsWith("00000"))
            {
                var positionChar = hash[5];
                if (char.IsDigit(positionChar))
                {
                    var position = positionChar - '0';
                    if (position < passwordLength && !password[position].HasValue)
                    {
                        password[position] = hash[6];
                        filledPositions++;
                    }
                }
            }

            index++;
        }

        return new string(password.Select(c => c ?? '_').ToArray());
    }

    public static void Run()
    {
        Console.WriteLine($"Part One: {SolvePartOne()}");
        Console.WriteLine($"Part Two: {SolvePartTwo()}");
    }
}

File: 2017/05.fs
namespace _2017

open System;

module _05 =
    let Data = (Utils.GetInputData 5).Split('\n', StringSplitOptions.RemoveEmptyEntries) |> Array.map int

    let executeWithRule (modifyJump: int -> int) =
        let mutable instructions = Array.copy Data
        let mutable index = 0
        let mutable steps = 0
        while index >= 0 && index < instructions.Length do
            let jump = instructions.[index]
            instructions.[index] <- modifyJump jump
            index <- index + jump
            steps <- steps + 1
        steps

    let solvePartOne() = executeWithRule (fun jump -> jump + 1)

    let solvePartTwo() = executeWithRule (fun jump -> if jump >= 3 then jump - 1 else jump + 1)


    let Run() =
        printfn $"Part One: {solvePartOne()}"
        printfn $"Part Two: {solvePartTwo()}"